testplan/common/utils/monitor/__init__.py

```diff
@@ -0,0 +1,295 @@
+        file_name = 'monitor-{}-{}.data'.format(self.hostname, self.uid)
+        full_path = os.path.join(directory, file_name)
+        if not os.path.exists(directory):
+            TESTPLAN_LOGGER.info('Creating directory %s', directory)
+            makedirs(directory)
+        with open(full_path, 'w') as monitor_file:
+            monitor_file.write(self.dumps())
+        TESTPLAN_LOGGER.debug('Monitor data saved to {}'.format(full_path))
+
+    def load(self, serial_json):
+        """
+        Load data from a JSON string or a dict.
+
+        :param serial_json: The json type of event monitor
+        :type serial_json: ``str`` or ``dict``
+
+        :return: ``None``
+        :rtype: ``NoneType``
+        """
+        if isinstance(serial_json, str):
+            _serial_json = json.loads(serial_json)
+        else:
+            _serial_json = serial_json
+        for name, value in _serial_json.items():
+            setattr(self, name, value)
+        TESTPLAN_LOGGER.info('Internal data loaded, previous internal data overwritten.')
+
+    def to_event_monitor(self):
+        return self
+
+
+class ResourceMonitor(EventMonitor):
+    """
+    Resource Monitor will collect the host hardware information(CPU, memory, disk etc.)
+    and the metrics(the usage of CPU, memory and disk).
+
+    """
+
+    def __init__(self, directory=None, parent=None, collector=None):
+        super(ResourceMonitor, self).__init__(parent=parent)
+        self._collector = collector or get_collector(directory)
+        self._poll = None
+        self._poll_event = threading.Event()
+        self._poll_start = False
+        self.poll_interval = 2
+        self.host_metadata = self._collector.metadata
+        self.monitor_metrics = {
+            'cpu': [],
+            'memory': [],
+            'disk': [],
+            'iops': [],
+            'read': [],
+            'write': [],
+            'time': []
+        }
+
+    def _monitoring(self):
+        while self._poll_start:
+            _now = time.time()
+            _metrics = self._collector.monitor()
+            _metrics['time'] = _now
+            for key in self.monitor_metrics.keys():
+                self.monitor_metrics[key].append(_metrics[key])
+
+            _sleep_time = self.poll_interval - (time.time() - _now)
+            if _sleep_time > 0:
+                time.sleep(_sleep_time)
+
+    def start(self):
+        """
+        Start the thread of collect the metrics.
+
+        :return: ``None``
+        :rtype: ``NoneType``
+        """
+        self._poll = threading.Thread(target=self._monitoring)
+        self._poll.daemon = True
+        self._poll_start = True
+        self._poll.start()
+        TESTPLAN_LOGGER.info('Resource polling started.')
+
+    def stop(self):
+        """
+        Stop collect metrics.
+
+        :return: ``None``
+        :rtype: ``NoneType``
+        """
+        if self._poll:
+            self._poll_start = False
```

Comment:
It's a deamon thread, so I don't think we need to join() it.

Reply:
You should always join() threads when they are exiting as a matter of good practice. Marking as a daemon is just to ensure that it doesn't block the interpreter from exiting in case we abort (e.g. due to KeyboardInterrupt), but for normal exit we wait for threads to terminate.